From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: UniFrog CI <ci@unifrog.local>
Date: Mon, 22 Jun 2026 07:19:09 +0000
Subject: [PATCH 2/2] Harden PICO-8 print control parsing

Validate every control-sequence operand before advancing through the string, bound emulated RAM writes, remove the unused fixed-size audio buffer, prevent zero-width tab modulo operations, and restore the solid-background disable selector.
---
 source/printHelper.cpp | 61 +++++++++++++++++++++++++++++-------------
 1 file changed, 43 insertions(+), 18 deletions(-)

diff --git a/source/printHelper.cpp b/source/printHelper.cpp
index 2b423a97067b0012ed2ff5cea132b5370cb8201b..aa4d4e0f9e435bc3d4e7b7df36f1910ab98f75e6 100644
--- a/source/printHelper.cpp
+++ b/source/printHelper.cpp
@@ -1,9 +1,9 @@
+#include <algorithm>
 #include <string>
 
 #include "printHelper.h"
 #include "nibblehelpers.h"
 
-#include <string>
 #include "graphics.h"
 #include "vm.h"
 #include "Audio.h"
@@ -133,7 +133,7 @@ int print(std::string str, int x, int y, uint8_t c) {
 
 	_ph_mem->drawState.text_x = x;
 	_ph_mem->drawState.text_y = y;
-    int length = str.length();
+    size_t length = str.length();
     int homeX = x;
     int homeY = y;
     int prevX = x;
@@ -147,7 +147,6 @@ int print(std::string str, int x, int y, uint8_t c) {
     uint8_t bgColor = 0xff;
     uint8_t fgColor = _ph_mem->drawState.color;
     uint8_t charBytes[8];
-    uint8_t audioStrBytes[32];
     bool cancelWrap = false;
 
     uint8_t printMode = _ph_mem->hwState.printAttributes;
@@ -178,6 +177,7 @@ int print(std::string str, int x, int y, uint8_t c) {
             return x;
         }
         else if (ch == 1) { // "\*{p0}" repeat the next character p0 times
+            if (n + 2 >= length) break;
 			uint8_t timesChr = str[++n];
 			int times = p0CharToNum(timesChr);
 
@@ -196,10 +196,12 @@ int print(std::string str, int x, int y, uint8_t c) {
 			}
 		}
 		else if (ch == 2) { // "\#{p0}" draw text on a solid background color
+            if (n + 1 >= length) break;
 			uint8_t bgColChar = str[++n];
 			bgColor = p0CharToNum(bgColChar);
 		}
 		else if (ch == 3) { // "\-{p0}" move text cursor horizontally by 16-p0 pixels
+            if (n + 1 >= length) break;
 			uint8_t pixelCountChar = str[++n];
 			int deltaX = p0CharToNum(pixelCountChar) - 16;
 
@@ -207,6 +209,7 @@ int print(std::string str, int x, int y, uint8_t c) {
 			//update memory cursor state?
 		}
 		else if (ch == 4) { // "\|{p0}" move text cursor vertically by 16-p0 pixels
+            if (n + 1 >= length) break;
 			uint8_t pixelCountChar = str[++n];
 			int deltaY = p0CharToNum(pixelCountChar) - 16;
 
@@ -214,6 +217,7 @@ int print(std::string str, int x, int y, uint8_t c) {
 			//update memory cursor state?
 		}
 		else if (ch == 5) { // "\+{p0}" move text cursor vertically by 16-p0 pixels
+            if (n + 2 >= length) break;
 			uint8_t pixelCountXChar = str[++n];
 			uint8_t pixelCountYChar = str[++n];
 			int deltaX = p0CharToNum(pixelCountXChar) - 16;
@@ -224,6 +228,7 @@ int print(std::string str, int x, int y, uint8_t c) {
 			//update memory cursor state?
 		}
 		else if (ch == 6) { // "\^" special command
+            if (n + 1 >= length) break;
 			uint8_t commandChar = str[++n];
             // \^1 through \^9 skips a number of frames before continuing printing
 			if (commandChar > 48 && commandChar < 58){
@@ -236,11 +241,13 @@ int print(std::string str, int x, int y, uint8_t c) {
                 }
 			}
             // \^d P0 causes subsequent printing to pause P0 frames between each character,
-            else if (commandChar == 'd') {
+			else if (commandChar == 'd') {
+                if (n + 1 >= length) break;
                 uint8_t framesChar = str[++n];
                 framesBetweenChars = p0CharToNum(framesChar);
             }
             else if (commandChar == 'c'){
+                if (n + 1 >= length) break;
                 uint8_t colChar = str[++n];
                 uint8_t col = p0CharToNum(colChar);
 
@@ -255,20 +262,25 @@ int print(std::string str, int x, int y, uint8_t c) {
                 homeY = y;
             }
             else if (commandChar == 'j'){
+                if (n + 2 >= length) break;
                 uint8_t xChar = str[++n];
                 uint8_t yChar = str[++n];
                 x = p0CharToNum(xChar) * 4;
                 y = p0CharToNum(yChar) * 4;
             }
             else if (commandChar == 'r'){
+                if (n + 1 >= length) break;
                 uint8_t rhsWrapChar = str[++n];
                 rhsWrap = p0CharToNum(rhsWrapChar) * 4;
             }
             else if (commandChar == 's'){
+                if (n + 1 >= length) break;
                 uint8_t tswChar = str[++n];
-                tabStopWidth = p0CharToNum(tswChar);
+                int width = p0CharToNum(tswChar);
+                tabStopWidth = width > 0 ? width : 1;
             }
             else if (commandChar == 'x'){
+                if (n + 1 >= length) break;
                 uint8_t charWidthChar = str[++n];
                 forceCharWidth = p0CharToNum(charWidthChar);
                 charWidth = forceCharWidth;
@@ -276,6 +288,7 @@ int print(std::string str, int x, int y, uint8_t c) {
             else if (commandChar == 'y'){
                 //this behaves in a way I wouldn't expect in pico 8- it seems to apply to the next print statment
                 //but not if there is a \n in the string. Not sure if this is a bug or not
+                if (n + 1 >= length) break;
                 uint8_t charHeightChar = str[++n];
                 forceCharHeight = p0CharToNum(charHeightChar);
                 charHeight = forceCharHeight;
@@ -315,6 +328,9 @@ int print(std::string str, int x, int y, uint8_t c) {
                 printMode |= PRINT_MODE_SOLID_BG;
             }
             else if (commandChar == ':' || commandChar == '.'){
+                size_t byteCount = commandChar == ':' ? 16 : 8;
+
+                if (byteCount > length - n - 1) break;
                 charHeight = forceCharHeight > 0 ? forceCharHeight : 8;
                 if (commandChar == ':') {
                     std::string hexStr = str.substr(n+1, 16);
@@ -344,6 +360,7 @@ int print(std::string str, int x, int y, uint8_t c) {
                 lineHeight = charHeight > lineHeight ? charHeight : lineHeight;
             }
             else if (commandChar == '-'){
+                if (n + 1 >= length) break;
                 uint8_t turnOffModeChar = str[++n];
                 if (printMode) {
                     if (turnOffModeChar == 'w') {
@@ -370,7 +387,7 @@ int print(std::string str, int x, int y, uint8_t c) {
                     else if (turnOffModeChar == 'b') {
                         printMode &= ~(PRINT_MODE_PADDING);
                     }
-                    else if (turnOffModeChar == 'b') {
+                    else if (turnOffModeChar == '#') {
                         printMode &= ~(PRINT_MODE_SOLID_BG);
                     }
                 }
@@ -379,46 +396,53 @@ int print(std::string str, int x, int y, uint8_t c) {
                 //no wraping after this print call (may be pico 8 bug?)
                 cancelWrap = true;
 
+                if (4 > length - n - 1) break;
                 std::string addrHexStr = str.substr(n+1, 4);
                 n+=4;
-                int addr = (int)strtol(addrHexStr.c_str(), NULL, 16);
+                size_t addr = (size_t)strtoul(addrHexStr.c_str(), NULL, 16);
 
-                int size = length - n;
+                size_t size = length - n - 1;
                 std::string binStr = str.substr(n + 1, size);
-                n+=size;
-
+                size_t available = addr < sizeof(_ph_mem->data) ?
+                    sizeof(_ph_mem->data) - addr : 0;
+                size = std::min(size, available);
                 for(size_t i = 0; i < size; i++) {
                     _ph_mem->data[addr + i] = binStr[i];
                 }
+                n = length - 1;
             }
             else if (commandChar == '@'){
+                if (8 > length - n - 1) break;
                 std::string addrHexStr = str.substr(n+1, 4);
                 n+=4;
-                int addr = (int)strtol(addrHexStr.c_str(), NULL, 16);
+                size_t addr = (size_t)strtoul(addrHexStr.c_str(), NULL, 16);
 
                 std::string sizeHexStr = str.substr(n+1, 4);
                 n+=4;
 
-                int size = (int)strtol(sizeHexStr.c_str(), NULL, 16);
+                size_t size = (size_t)strtoul(sizeHexStr.c_str(), NULL, 16);
+                if (size > length - n - 1) break;
 
                 std::string binStr = str.substr(n + 1, size);
                 n+= size;
 
+                size_t available = addr < sizeof(_ph_mem->data) ?
+                    sizeof(_ph_mem->data) - addr : 0;
+                size = std::min(size, available);
                 for(size_t i = 0; i < size; i++) {
                     _ph_mem->data[addr + i] = binStr[i];
                 }
             }
 		}
 		else if (ch == 7) { // "\a" audio command
-            uint8_t nextChar = str[++n];
-            int audioStrIdx = 0;
-            while (nextChar != 32 && n < length) {
-                audioStrBytes[audioStrIdx++] = nextChar;
-                nextChar = str[++n];
-            }
+            while (n + 1 < length && str[n + 1] != 32)
+                n++;
+            if (n + 1 < length)
+                n++;
             //todo: generate audio stuff and play
         }
         else if (ch == 11) { // "\v" decorate previous character
+            if (n + 2 >= length) break;
             uint8_t offsetChar = str[++n];
 			uint8_t toPrint = str[++n];
             int offset = p0CharToNum(offsetChar);
@@ -437,6 +461,7 @@ int print(std::string str, int x, int y, uint8_t c) {
                 forceCharHeight);
         }
         else if (ch == 12) { //"\f{p0}" draw text with this foreground color
+            if (n + 1 >= length) break;
 			uint8_t fgColChar = str[++n];
 			fgColor = p0CharToNum(fgColChar);
 			_ph_graphics->color(fgColor);
-- 
2.54.0

